psl-core/src/main/java/org/linqs/psl/reasoner/sgd/SGDReasoner.java

```diff
@@ -60,65 +78,86 @@ public double optimize(TermStore baseTermStore) {
 
         termStore.initForOptimization();
 
-        double objective = -1.0;
+        float meanMovement = 0.0f;
+        long termCount = 0;
+        double objective = Double.POSITIVE_INFINITY;
         double oldObjective = Double.POSITIVE_INFINITY;
 
+        HashMap<Integer, Float> accumulatedGradientSquares = null;
+        HashMap<Integer, Float> accumulatedGradientMean = null;
+        HashMap<Integer, Float>  accumulatedGradientVariance = null;
+
+        // Initialize dynamic data structures for optimization.
+        float[] oldVariableValues = null;
+        if (adaGrad) {
+            accumulatedGradientSquares = new HashMap<Integer, Float>();
+        } else if (adam) {
+            accumulatedGradientMean = new HashMap<Integer, Float>();
+            accumulatedGradientVariance = new HashMap<Integer, Float>();
+        }
+
         if (printInitialObj && log.isTraceEnabled()) {
             objective = computeObjective(termStore);
-            log.trace("Iteration {} -- Objective: {}, Mean Movement: {}, Iteration Time: {}, Total Optimiztion Time: {}", 0, objective, 0.0f, 0, 0);
+            log.trace("Iteration {} -- Objective: {}, Mean Movement: {}, Iteration Time: {}, Total Optimization Time: {}", 0, objective, 0.0f, 0, 0);
         }
 
-        int iteration = 1;
         long totalTime = 0;
-        while (true) {
+        boolean converged = false;
+        for (int iteration = 1; iteration < (maxIterations * budget) && !converged; iteration++) {
             long start = System.currentTimeMillis();
 
-            // Keep track of the mean movement of the random variables.
-            float movement = 0.0f;
+            termCount = 0;
+            meanMovement = 0.0f;
+            objective = 0.0;
 
-            float[] variableValues = termStore.getVariableValues();
             for (SGDObjectiveTerm term : termStore) {
-                movement += term.minimize(iteration, variableValues);
+                if (oldVariableValues != null) {
```

Comment:
Again similar comment, can you set oldVariableValues to the first term? Will it ever be null again? Also this pattern is not the same as DCD. Do you want to be running if oldVariableValues is null?